### PR TITLE
force k8s-infra-prow-build jobs to be Guaranteed Pod QOS

### PR DESCRIPTION
**Pull request by spiffxp**
Add presubmit to enforce a policy that any jobs scheduled for k8s-infra-prow-builds must have Guaranteed Pod QOS, meaning they must have matching non-zero cpu and memory resource limits and requests.

Add guesses at resource limits for all jobs currently scheduled for k8s-infra-prow-builds.  The guesses are based off of visual inspection of memory and CPU usage on the cluster using GCP Monitoring's Metrics Explorer.  (If you can't see these links, you'll need to be a member of [k8s-infra-prow-viewers@kubernetes.io](https://github.com/kubernetes/k8s.io/blob/8e267aa1ed12b26897dbf9bb2d6c1294cdd76dbd/groups/groups.yaml#L679-L688), please PR yourself in if you'd like access):
- e.g. [memory usage for ci-kubernetes-kind-ipv2-e2e-parallel.*](https://console.cloud.google.com/monitoring/metrics-explorer?project=k8s-infra-prow-build&timeDomain=6h&pageState=%7B%22xyChart%22:%7B%22dataSets%22:%5B%7B%22timeSeriesFilter%22:%7B%22filter%22:%22metric.type%3D%5C%22kubernetes.io%2Fcontainer%2Fmemory%2Fused_bytes%5C%22%20resource.type%3D%5C%22k8s_container%5C%22%20metadata.user_labels.%5C%22prow.k8s.io%2Fjob%5C%22%3Dmonitoring.regex.full_match(%5C%22ci-kubernetes-kind-ipv6-e2e-parallel.*%5C%22)%22,%22minAlignmentPeriod%22:%2260s%22,%22unitOverride%22:%22By%22,%22aggregations%22:%5B%7B%22perSeriesAligner%22:%22ALIGN_MEAN%22,%22crossSeriesReducer%22:%22REDUCE_SUM%22,%22groupByFields%22:%5B%22metadata.user_labels.%5C%22prow.k8s.io%2Fjob%5C%22%22%5D%7D,%7B%22crossSeriesReducer%22:%22REDUCE_NONE%22%7D%5D%7D,%22targetAxis%22:%22Y1%22,%22plotType%22:%22LINE%22%7D%5D,%22options%22:%7B%22mode%22:%22COLOR%22%7D,%22constantLines%22:%5B%5D,%22timeshiftDuration%22:%220s%22,%22y1Axis%22:%7B%22label%22:%22y1Axis%22,%22scale%22:%22LINEAR%22%7D%7D,%22isAutoRefresh%22:true,%22timeSelection%22:%7B%22timeRange%22:%226h%22%7D%7D)
- e.g. [CPU usage time for ci-kubernetes-gci-gce-scalability.*](https://console.cloud.google.com/monitoring/metrics-explorer?project=k8s-infra-prow-build&timeDomain=1h&pageState=%7B%22xyChart%22:%7B%22dataSets%22:%5B%7B%22timeSeriesFilter%22:%7B%22filter%22:%22metric.type%3D%5C%22kubernetes.io%2Fcontainer%2Fcpu%2Fcore_usage_time%5C%22%20resource.type%3D%5C%22k8s_container%5C%22%20metadata.user_labels.%5C%22prow.k8s.io%2Fjob%5C%22%3Dmonitoring.regex.full_match(%5C%22ci-kubernetes-e2e-gci-gce-scalability.*%5C%22)%22,%22minAlignmentPeriod%22:%2260s%22,%22unitOverride%22:%22s%22,%22aggregations%22:%5B%7B%22perSeriesAligner%22:%22ALIGN_RATE%22,%22crossSeriesReducer%22:%22REDUCE_SUM%22,%22groupByFields%22:%5B%22metadata.user_labels.%5C%22prow.k8s.io%2Fjob%5C%22%22%5D%7D,%7B%22crossSeriesReducer%22:%22REDUCE_NONE%22%7D%5D%7D,%22targetAxis%22:%22Y1%22,%22plotType%22:%22LINE%22%7D%5D,%22options%22:%7B%22mode%22:%22COLOR%22%7D,%22constantLines%22:%5B%5D,%22timeshiftDuration%22:%220s%22,%22y1Axis%22:%7B%22label%22:%22y1Axis%22,%22scale%22:%22LINEAR%22%7D%7D,%22isAutoRefresh%22:true,%22timeSelection%22:%7B%22timeRange%22:%221d%22%7D%7D)

I erred on the side of giving some overhead for memory (avoid oomkill), and avoiding overhead for CPU spikes (some light throttling shouldn't be the end of the world)

Add a test to compare the total limits of jobs scheduled for k8s-infra-prow-build against what its current min-nodes capacity is.  It looked like we would ask for more capacity than the cluster offered with its minimum number of nodes, so I preemptively doubled the number of nodes.

The k8s-infra-prow-build cluster currently runs no presubmits, only periodics, and almost entirely release-blocking or release-informing jobs (ref: https://github.com/kubernetes/k8s.io/issues/841).  I would be interested to see if this stabilizes jobs randomly flaking over there.  If so, we could use this as a model for enforcing well behaved presubmits.

### CODE REVIEW

**Comment by liggitt** [comment]
s/burstable/guaranteed/?

**Comment by spiffxp** [comment]
>s/burstable/guaranteed/?

I pushed what I think that would look like.  I was using burstable to figure out which jobs lacked any resource requests at all

**Comment by liggitt** [comment]
ah

**Comment by spiffxp** [comment]
/hold
for comment

File: config/jobs/kubernetes/sig-scalability/sig-scalability-release-blocking-jobs.yaml [hunk 1]
```diff
@@ -168,3 +168,10 @@ periodics:
       - --test-cmd-name=ClusterLoaderV2
       - --timeout=120m
       - --use-logexporter
+      resources:
+        requests:
+          cpu: 2
+          memory: 6Gi
```
Comment (wojtek-t) [inline config/jobs/kubernetes/sig-scalability/sig-scalability-release-blocking-jobs.yaml:174]: @tosi3k - you were running some experiments here. I think the outcome was that 2 CPUs were more than enough for 100-node tests, right?

Comment (liggitt) [reply config/jobs/kubernetes/sig-scalability/sig-scalability-release-blocking-jobs.yaml:174]: > @tosi3k - you were running some experiments here. I think the outcome was that 2 CPUs were more than enough for 100-node tests, right?

did you run the experiments with requests and limits, or just requests?

Comment (tosi3k) [reply config/jobs/kubernetes/sig-scalability/sig-scalability-release-blocking-jobs.yaml:174]: Just the requests w/o limits and now I see it was wrong... my bad.

Comment (spiffxp) [reply config/jobs/kubernetes/sig-scalability/sig-scalability-release-blocking-jobs.yaml:174]: Like I said in description, mine was a visual guess.  I think the actual peak I saw was somewhere around 1200m-1500m but I wanted to stick to whole numbers of CPUs for this first pass.  If you want to open a PR against https://github.com/kubernetes/k8s.io/blob/8e267aa1ed12b26897dbf9bb2d6c1294cdd76dbd/groups/groups.yaml#L679-L688 to add yourself to the group and see what I saw I'm happy to merge.

**Comment by spiffxp** [comment]
FYI @kubernetes/ci-signal @kubernetes/release-engineering

**Comment by spiffxp** [comment]
I would like to hear from @kubernetes/ci-signal and @onlydole if they're OK with this rolling out.  It may result in some of the -blocking boards being bumpier for a bit while we figure out the appropriate values.  However, I suspect the feedback cycle will be pretty quick.

**Comment by jimangel** [comment]
For my own understanding, is there a difference in requesting 1000m/2000m instead of 1/2 cpus?

**Comment by BenTheElder** [comment]
> For my own understanding, is there a difference in requesting 1000m/2000m instead of 1/2 cpus?

nope, it's all resource.Quanity values
https://github.com/kubernetes/apimachinery/blob/945d4ebf362b3bbbc070e89371e69f9394737676/pkg/api/resource/quantity.go#L30

**Comment by onlydole** [comment]
@spiffxp - I am in **full** support of having non-zero cpu and memory resource limits and requests going forward. I see that as a big win for our testing suite as a whole, and in making releases more sustainable in the long-term.

**Comment by BenTheElder** [comment]
I don't think we have enough RAM on any node available to schedule verify https://k8s-testgrid.appspot.com/sig-release-master-blocking#verify-master

**Comment by BenTheElder** [comment]
scratch that: we have nodes that the requests should fit on, but it's mysteriously failing reliably, starting after this merged.

a node:
451 mCPU | 7.91 CPU | 352.32 MB | 49.7 GB

**Comment by BenTheElder** [comment]
it is in fact failing due to pod scheduling timeout though.

**Comment by BenTheElder** [comment]
This is very frustrating to debug. I can't find anything useful in the logs (incl. stackdriver) or the artifacts, just the prowjob failure message that the pod failed scheduling. No details on why (memory? cpu?)

**Comment by BenTheElder** [comment]
https://github.com/kubernetes/test-infra/issues/18530

**Pull request merged**